### PR TITLE
[Update] v1.0.2b

### PR DESCRIPTION
**Pull request by TokyoSU**
hi @TeslaRus,
changelog here: [Releases](https://github.com/TokyoSU/OpenTomb/releases)
ps: need to change cmakelist for weapons.cpp,
i'm not using cmake so cmakelists is not included here !

### CODE REVIEW

File: scripts/level/tr1/LEVEL6.lua [hunk 1]
```diff
@@ -6,6 +6,7 @@ level_PostLoad = function()
     moveEntityLocal(player, 0, 0, 256);
     setEntityMoveType(player, MOVE_UNDERWATER);
     setEntityAnim(player, ANIM_TYPE_BASE, 108, 0);
+    rewriteName(player, ITEM_PUZZLE_1, ITEM_NAME_TR1_PUZZLE_GOLD_BAR);
```
Comment (TokyoSU) [inline scripts/level/tr1/LEVEL6.lua]: not found the iron bar :x

Comment (TeslaRus) [reply scripts/level/tr1/LEVEL6.lua]: iron bar id == 100

**Comment by TeslaRus** [comment]
okay... I need some time to review it ;-)

File: scripts/level/tr1/LEVEL6.lua (hunk 1 above)
Comment (TokyoSU) [reply scripts/level/tr1/LEVEL6.lua]: thanks :)

File: src/gui/gui.cpp [hunk 2]
```diff
@@ -543,18 +550,20 @@ bool Gui_LoadScreenAssignPic(const char* pic_name)
     size_t pic_len = strlen(pic_name);
     size_t base_len = strlen(Engine_GetBasePath());
     size_t buf_len = pic_len + base_len + 5;
-    char image_name_buf[buf_len];
+    char *image_name_buf = new char[buf_len];
```
Comment (TokyoSU) [inline src/gui/gui.cpp]: cause error in windows.

File: src/gui/gui_inventory.h [hunk 3]
```diff
@@ -115,17 +117,20 @@ enum inventoryState
...
     void render();
-    uint32_t getItemIdActualView();
 
     gl_text_line_t              m_label_title;
     char                        m_label_title_text[GUI_LINE_DEFAULTSIZE];
     gl_text_line_t              m_label_item_name;
     char                        m_label_item_name_text[GUI_LINE_DEFAULTSIZE];
 
+    // get item_id by view or selected !
+    uint32_t getItemIdActualView();
```
Comment (TokyoSU) [inline src/gui/gui_inventory.h]: now used, sorry for missed using at the preview update.

File: scripts/level/tr1/LEVEL6.lua (hunk 1 above)
Comment (TokyoSU) [reply scripts/level/tr1/LEVEL6.lua]: #added

**Comment by TokyoSU** [comment]
the #c3f200e / #3149b84 and #7604359 are fail, sorry for it :x

File: src/weapons.h [hunk 4]
```diff
@@ -0,0 +1,382 @@
...
+{
+    uint32_t item_id;
+    int ammo_counter;
+    int shot;
+    int draw;
+    int hide;
+    int echo;
+    int reload_1;     // classic reload
+    int reload_2;     // grenadegun reload lock
+    int damage = NULL;
```
Comment (RICCIARDI-Adrien) [inline src/weapons.h]: The 3 preceding fields are not pointers, they should not be initialized using NULL (this triggers a warning when building with gcc).
```suggestion
    int damage = 0;
```

File: src/weapons.h [hunk 5]
```diff
@@ -0,0 +1,382 @@
...
+    int damage_water = NULL;
```
Comment (RICCIARDI-Adrien) [inline src/weapons.h]:
```suggestion
    int damage_water = 0;
```

File: src/weapons.h [hunk 6]
```diff
@@ -0,0 +1,382 @@
...
+    int damage_explosion = NULL;
```
Comment (RICCIARDI-Adrien) [inline src/weapons.h]:
```suggestion
    int damage_explosion = 0;
```

File: src/weapons.h [hunk 7]
```diff
@@ -0,0 +1,382 @@
...
+    SND_M16_STOP = TR_AUDIO_SOUND_SHOTM16_END,
+    SND_MP5 = TR_AUDIO_SOUND_SHOTMP5,
+    SND_ROCKETGUN = TR_AUDIO_SOUND_SHOTROCKETGUN,
+    SND_ROCKETGUN_RELOAD = TR_AUDIO_SOUND_RELOADROCKETGUN,
+    SND_HARPOON_LAND = TR_AUDIO_SOUND_SHOTHARPOON_G,
+    SND_HARPOON_WATER = TR_AUDIO_SOUND_SHOTHARPOON_W,
+    SND_HARPOON_RELOAD_LAND = TR_AUDIO_SOUND_RELOADHARPOON_G,
+    SND_HARPOON_RELOAD_WATER = TR_AUDIO_SOUND_RELOADHARPOON_W,
+    SND_GRENADEGUN = TR_AUDIO_SOUND_SHOTGRENADEGUN,
+    SND_TR2_3_GRENADEGUN_RELOAD = TR_AUDIO_SOUND_TR2_3_RELOADGRENADEGUN,
```
Comment (RICCIARDI-Adrien) [inline src/weapons.h]: I did not find `TR_AUDIO_SOUND_TR2_3_RELOADGRENADEGUN` declaration in the code (is it commited ?) :
```
In file included from /home/ar/Documents/OpenTombTokyoSU/src/gui/gui_inventory.cpp:30:
/home/ar/Documents/OpenTombTokyoSU/src/gui/../weapons.h:221:35: error: ‘TR_AUDIO_SOUND_TR2_3_RELOADGRENADEGUN’ was not declared in this scope
```

File: src/weapons.h [hunk 8]
```diff
@@ -0,0 +1,382 @@
...
+    SND_TR2_3_GRENADEGUN_LOCK = TR_AUDIO_SOUND_TR2_3_RELOADGRENADEGUN_LOCK,
```
Comment (RICCIARDI-Adrien) [inline src/weapons.h]:
```
/home/ar/Documents/OpenTombTokyoSU/src/gui/../weapons.h:222:33: error: ‘TR_AUDIO_SOUND_TR2_3_RELOADGRENADEGUN_LOCK’ was not declared in this scope
```

File: src/weapons.h [hunk 9]
```diff
@@ -0,0 +1,382 @@
...
+    SND_TR4_C_GRENADEGUN_RELOAD = TR_AUDIO_SOUND_TR4C_RELOADGRENADEGUN,
```
Comment (RICCIARDI-Adrien) [inline src/weapons.h]:
```
/home/ar/Documents/OpenTombTokyoSU/src/gui/../weapons.h:223:35: error: ‘TR_AUDIO_SOUND_TR4C_RELOADGRENADEGUN’ was not declared in this scope
```

File: src/weapons.h [hunk 10]
```diff
@@ -0,0 +1,382 @@
...
+    SND_TR4_C_GRENADEGUN_LOCK = TR_AUDIO_SOUND_TR4C_RELOADGRENADEGUN_LOCK,
```
Comment (RICCIARDI-Adrien) [inline src/weapons.h]:
```
/home/ar/Documents/OpenTombTokyoSU/src/gui/../weapons.h:224:33: error: ‘TR_AUDIO_SOUND_TR4C_RELOADGRENADEGUN_LOCK’ was not declared in this scope
```

File: src/gui/gui.cpp [hunk 11]
```diff
@@ -30,6 +30,7 @@
 #include "gui.h"
 #include "gui_menu.h"
 #include "gui_inventory.h"
+#include "inventory.h"
```
Comment (RICCIARDI-Adrien) [inline src/gui/gui.cpp]:
```
/home/ar/Documents/OpenTombTokyoSU/src/gui/gui.cpp:33:10: fatal error: inventory.h: Aucun fichier ou dossier de ce type
```
```suggestion
#include "../inventory.h"
```

Comment (TokyoSU) [reply src/gui/gui.cpp]: #fixed thanks :)

File: src/weapons.h (hunk 10 above)
Comment (TokyoSU) [reply src/weapons.h]: #fixed missing audio.h

File: src/weapons.h (hunk 7 above)
Comment (TokyoSU) [reply src/weapons.h]: #fixed missing audio.h

File: src/weapons.h (hunk 6 above)
Comment (TokyoSU) [reply src/weapons.h]: #fixed thanks for reply

File: src/weapons.h (hunk 5 above)
Comment (TokyoSU) [reply src/weapons.h]: #fixed

File: src/weapons.h (hunk 4 above)
Comment (TokyoSU) [reply src/weapons.h]: #fixed

**Comment by TokyoSU** [comment]
thanks you @RICCIARDI-Adrien for reviewing, the last commit, fix all review.

File: src/state_control/state_control_Lara.cpp [hunk 12]
```diff
@@ -3279,144 +3275,84 @@ int StateControl_LaraDoOneHandWeaponFrame(struct entity_s *ent, struct  ss_anima
...
+        static float d_to[3] = { 0.0f, 0.0f, 0.0f };
+        static float color[4] = { 1.0f, 0.0f, 0.0f, 1.0f };
+        renderer.debugDrawer->DrawLine(d_from, d_to, color, color);*/
+
         entity_p target = (ent->character->target_id != ENTITY_ID_NONE) ? World_GetEntityByID(ent->character->target_id) : (NULL);
         ss_bone_tag_p b_tag = ent->bf->bone_tags + ent->character->bone_torso;
         bool do_aim = ent->character->cmd.action;
         float target_pos[3];
-        int inc_state;
+        int inc_state = NULL;
```
Comment (RICCIARDI-Adrien) [inline src/state_control/state_control_Lara.cpp]:
```suggestion
        int inc_state = 0;
```

**Comment by RICCIARDI-Adrien** [comment]
You're welcome @TokyoSU.
Here is the patch to add `weapons.*` files to the CMake build system :
```
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 518dd8b6..69e4380c 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -208,6 +208,8 @@ set(OPENTOMB_SRCS
     src/skeletal_model.c
     src/trigger.cpp
     src/trigger.h
+    src/weapons.cpp
+    src/weapons.h
     src/world.cpp
     src/world.h
 )
```

**Comment by RICCIARDI-Adrien** [comment]
There is a sound problem in TR2 : uzis firing sound does not stop playing.

File: src/weapons.cpp [hunk 13]
```diff
@@ -0,0 +1,1526 @@
...
+            return ITEM_MP5;  // it's h&k gun !
+        }
+        else if (GET_MODEL(TR4C_MODEL_REVOLVER))
+        {
+            return ITEM_REVOLVER;
+        }
+    }
+
+    // no id found
+    return NULL;
```
Comment (RICCIARDI-Adrien) [inline src/weapons.cpp]:
```suggestion
    return 0;
```
Is 0 an invalid ID (I did not dig into IDs code, it's just out of curiosity) ?

File: src/weapons.cpp [hunk 14]
```diff
@@ -0,0 +1,1526 @@
...
+    weapons[5] = deserteagle;
+    weapons[6] = revolver;
+    weapons[7] = m16;
+    weapons[8] = mp5;
+    weapons[9] = rocketgun;
+    weapons[10] = harpoongun;
+    weapons[11] = grenadegun;
+    weapons[12] = crossbowgun;
+    weapons[13] = grapplingun;
+    return weapons;
```
Comment (RICCIARDI-Adrien) [inline src/weapons.cpp]: Be careful, `weapons` is a local variable. This function won't work if called.

**Comment by TokyoSU** [comment]
i already know for uzi sound but this bug is also for m16 :x i will fix it

File: src/weapons.cpp [hunk 15]
```diff
@@ -0,0 +1,1526 @@
...
+    grapplingun.current_ammo = AMMO_GRAPPLIN;
+    grapplingun.onWater = false;
+    grapplingun.dealDmgAtImpact = false;
+    grapplingun.haveGravity = false;
+    // no muzzleflash for this ?
+    grapplingun.muzzle_pos[0] = NULL;
+    grapplingun.muzzle_pos[1] = NULL;
+    grapplingun.muzzle_pos[2] = NULL;
+    grapplingun.muzzle_orient[0] = NULL;
+    grapplingun.muzzle_orient[1] = NULL;
```
[the remark this answers is not in the document]
Comment (TokyoSU) [reply src/weapons.cpp]: #fixed

File: src/weapons.cpp [hunk 16]
```diff
@@ -0,0 +1,1526 @@
...
+    grapplingun.muzzle_orient[2] = NULL;
```
[the remark this answers is not in the document]
Comment (TokyoSU) [reply src/weapons.cpp]: #fixed

File: src/weapons.cpp (hunk 13 above)
Comment (TokyoSU) [reply src/weapons.cpp]: 0 is null because 1 is for pistol #fixed

File: src/weapons.cpp [hunk 17]
```diff
@@ -0,0 +1,1526 @@
...
+    grapplingun.bullet = 1;
+    grapplingun.current_ammo = AMMO_GRAPPLIN;
+    grapplingun.onWater = false;
+    grapplingun.dealDmgAtImpact = false;
+    grapplingun.haveGravity = false;
+    // no muzzleflash for this ?
+    grapplingun.muzzle_pos[0] = NULL;
+    grapplingun.muzzle_pos[1] = NULL;
+    grapplingun.muzzle_pos[2] = NULL;
+    grapplingun.muzzle_orient[0] = NULL;
```
[the remark this answers is not in the document]
Comment (TokyoSU) [reply src/weapons.cpp]: #fixed

File: src/weapons.cpp [hunk 18]
```diff
@@ -0,0 +1,1526 @@
...
+    grapplingun.firerate = RATE_GRAPPINGUN;
+    grapplingun.bullet = 1;
+    grapplingun.current_ammo = AMMO_GRAPPLIN;
+    grapplingun.onWater = false;
+    grapplingun.dealDmgAtImpact = false;
+    grapplingun.haveGravity = false;
+    // no muzzleflash for this ?
+    grapplingun.muzzle_pos[0] = NULL;
+    grapplingun.muzzle_pos[1] = NULL;
+    grapplingun.muzzle_pos[2] = NULL;
```
[the remark this answers is not in the document]
Comment (TokyoSU) [reply src/weapons.cpp]: #fixed

File: src/weapons.cpp [hunk 19]
```diff
@@ -0,0 +1,1526 @@
...
+    grapplingun.damage = 0;
+    grapplingun.firerate = RATE_GRAPPINGUN;
+    grapplingun.bullet = 1;
+    grapplingun.current_ammo = AMMO_GRAPPLIN;
+    grapplingun.onWater = false;
+    grapplingun.dealDmgAtImpact = false;
+    grapplingun.haveGravity = false;
+    // no muzzleflash for this ?
+    grapplingun.muzzle_pos[0] = NULL;
+    grapplingun.muzzle_pos[1] = NULL;
```
[the remark this answers is not in the document]
Comment (TokyoSU) [reply src/weapons.cpp]: #fixed

File: src/weapons.cpp [hunk 20]
```diff
@@ -0,0 +1,1526 @@
...
+    pistol.damage_explosion = 0;
+    pistol.firerate = RATE_PISTOL;
+    pistol.bullet = 2;
+    pistol.current_ammo = AMMO_UNLIMITED;
+    pistol.range = RANGE_PISTOL;
+    pistol.onWater = false;
+    pistol.dealDmgAtImpact = false;
+    pistol.haveGravity = false;
+    pistol.muzzle_duration = 3.0f;
+    pistol.muzzle_pos[0] = NULL;
```
[the remark this answers is not in the document]
Comment (TokyoSU) [reply src/weapons.cpp]: #fixed

File: src/weapons.cpp [hunk 21]
```diff
@@ -0,0 +1,1526 @@
...
+    pistol.muzzle_pos[1] = NULL;
```
[the remark this answers is not in the document]
Comment (TokyoSU) [reply src/weapons.cpp]: #fixed

File: src/weapons.cpp [hunk 22]
```diff
@@ -0,0 +1,1526 @@
...
+    pistol.muzzle_pos[2] = NULL;
```
[the remark this answers is not in the document]
Comment (TokyoSU) [reply src/weapons.cpp]: #fixed

File: src/weapons.cpp [hunk 23]
```diff
@@ -0,0 +1,1526 @@
...
+    pistol.muzzle_orient[0] = NULL;
```
[the remark this answers is not in the document]
Comment (TokyoSU) [reply src/weapons.cpp]: #fixed

File: src/weapons.cpp [hunk 24]
```diff
@@ -0,0 +1,1526 @@
...
+    pistol.muzzle_orient[1] = NULL;
```
[the remark this answers is not in the document]
Comment (TokyoSU) [reply src/weapons.cpp]: #fixed

File: src/weapons.cpp (hunk 14 above)
Comment (TokyoSU) [reply src/weapons.cpp]: #delete_and_fixed

File: src/state_control/state_control_Lara.cpp (hunk 12 above)
Comment (TokyoSU) [reply src/state_control/state_control_Lara.cpp]: #fixed

**Comment by TokyoSU** [comment]
> 
> 
> You're welcome @TokyoSU.
> Here is the patch to add `weapons.*` files to the CMake build system :
> 
> ```
> diff --git a/CMakeLists.txt b/CMakeLists.txt
> index 518dd8b6..69e4380c 100644
> --- a/CMakeLists.txt
> +++ b/CMakeLists.txt
> @@ -208,6 +208,8 @@ set(OPENTOMB_SRCS
>      src/skeletal_model.c
>      src/trigger.cpp
>      src/trigger.h
> +    src/weapons.cpp
> +    src/weapons.h
>      src/world.cpp
>      src/world.h
>  )
> ```

i can just push the CMakeLists ? i not familiar with git :x

**Comment by TeslaRus** [comment]
state control and below will be reviewed later; thanks for PR ;-)
P.S. I have to debug changes before merge too - some extra time...

**Comment by TokyoSU** [comment]
need time too, need to commit some files before merging.

**Comment by TokyoSU** [comment]
i want to use the "ss_anim->anim_frame_flags == ANIM_FRAME_REVERSE" but it seems that not working for unknown reason.
how to use this flag ? i use it with:
"ent->character->state.weapon_ready && !ent->character->cmd.action" for more precision.
is for uzi and m16 loop fix, but i wan't to fix echo not playing.
With (-time) echo sound are working but sound are playing when starting firing. :x

**Comment by TokyoSU** [comment]
update 1.0.2b changelog:
- fixed exit grenadegun animation (need to fix firing anim).
- fixed uzi loop sound.
- (fixed?) revolver and crossbow lasersight in inventory are now hidden. (can be unhide with new variable "itemIsEquipped" in "weapons_s".
screen_20190220_135250
/for in-game (model on lara anim) need to be changed when lasersight is equiped !.
- reverted weapons.cpp, now this class dont use define for function (but variables use define).
shotgun_old
- organizing weapons.h to be more clean.
- getVersion() now have more id for checking level version.
version
- now hp bar will only draw in tr1 to tr3 like original games !

**Comment by TokyoSU** [comment]
flares can be considered as a weapon ?
just need to create a dynamic light and a timer to switch off the light and launch the flare.
but for pickup the flare after launched (if light still exists), i dont know how to set up :x

**Comment by TokyoSU** [comment]
save_path need to be created if not found or the game.cpp throw warning that save can't be created !

**Comment by TokyoSU** [comment]
i restructured state folder:
state_new_structuration
in tr1 folder:
tr1_state_new_structuration
sure i change all #import and cmakelists.txt for you but i need your approval for it because state_control folder in opentomb repo need to be cleared before/after PR is pushed.
this restructuring of state_control folder is more clean, and distributed in multiple folder and name is lowered. this commits is for an another PR.

PS: tr2 entity now have .lua files but no function for now, tiger in test:
tr2_entity

**Comment by TeslaRus** [comment]
screen_20190304_221150
Target flag cleaning was lost

**Comment by TeslaRus** [comment]
resolved manually.
work in progress.
critical bugs must be fixed.

**Pull request merged**